lib/compiler/src/artifact_builders/artifact_builder.rs

```diff
@@ -242,3 +254,227 @@ impl ArtifactCreate for ArtifactBuild {
+        &self
+            .cell
+            .borrow_dependent()
+            .compilation
+            .function_relocations
+    }
+
+    /// Get Function Relocations ref
+    pub fn get_custom_section_relocations_ref(
+        &self,
+    ) -> &ArchivedPrimaryMap<SectionIndex, Vec<Relocation>> {
+        &self
+            .cell
+            .borrow_dependent()
+            .compilation
+            .custom_section_relocations
+    }
+
+    /// Get LibCall Trampoline Section Index
+    pub fn get_libcall_trampolines(&self) -> SectionIndex {
+        self.cell.borrow_dependent().compilation.libcall_trampolines
+    }
+
+    /// Get LibCall Trampoline Length
+    pub fn get_libcall_trampoline_len(&self) -> usize {
+        self.cell
+            .borrow_dependent()
+            .compilation
+            .libcall_trampoline_len as usize
+    }
+
+    /// Get Debug optional Dwarf ref
+    pub fn get_debug_ref(&self) -> Option<&Dwarf> {
+        match self.cell.borrow_dependent().compilation.debug {
+            ArchivedOption::Some(ref x) => Some(x),
+            ArchivedOption::None => None,
+        }
+    }
+
+    /// Get Function Relocations ref
+    pub fn deserialize_frame_info_ref(
+        &self,
+    ) -> Result<PrimaryMap<LocalFunctionIndex, CompiledFunctionFrameInfo>, DeserializeError> {
+        let mut deserializer = SharedDeserializeMap::new();
+        rkyv::Deserialize::deserialize(
+            &self.cell.borrow_dependent().compilation.function_frame_info,
+            &mut deserializer,
+        )
+        .map_err(|e| DeserializeError::CorruptedBinary(format!("{:?}", e)))
+    }
+}
+
+impl<'a> ArtifactCreate<'a> for ArtifactBuildFromArchive {
+    type OwnedDataInitializer = &'a ArchivedOwnedDataInitializer;
+    type OwnedDataInitializerIterator = core::slice::Iter<'a, ArchivedOwnedDataInitializer>;
+
+    fn create_module_info(&self) -> Arc<ModuleInfo> {
+        self.compile_info.module.clone()
+    }
+
+    fn set_module_info_name(&mut self, name: String) -> bool {
+        Arc::get_mut(&mut self.compile_info.module).map_or(false, |mut module_info| {
+            module_info.name = Some(name.to_string());
+            true
+        })
+    }
+
+    fn module_info(&self) -> &ModuleInfo {
+        &self.compile_info.module
+    }
+
+    fn features(&self) -> &Features {
+        &self.compile_info.features
+    }
+
+    fn cpu_features(&self) -> EnumSet<CpuFeature> {
+        EnumSet::from_u64(self.cell.borrow_dependent().cpu_features)
+    }
+
+    fn data_initializers(&'a self) -> Self::OwnedDataInitializerIterator {
+        self.cell.borrow_dependent().data_initializers.iter()
+    }
+
+    fn memory_styles(&self) -> &PrimaryMap<MemoryIndex, MemoryStyle> {
+        &self.compile_info.memory_styles
+    }
+
+    fn table_styles(&self) -> &PrimaryMap<TableIndex, TableStyle> {
+        &self.compile_info.table_styles
+    }
+
+    fn serialize(&self) -> Result<Vec<u8>, SerializeError> {
```

Comment:
I'm concerned on having some of this logic duplicated on both Plain and Archive... is it currently duplicated or did I get it wrong?

Reply:
The logic isn't exactly the same aside from the header, but I could easily refactor that bit out into a new function.